In [1]:
pip install pandas

    100% |████████████████████████████████| 9.5MB 160kB/s ta 0:00:011
    100% |████████████████████████████████| 13.4MB 126kB/s ta 0:00:011
    100% |████████████████████████████████| 512kB 2.4MB/s ta 0:00:011
You are using pip version 9.0.3, however version 21.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install scikit-learn

    100% |████████████████████████████████| 20.0MB 91kB/s eta 0:00:0111    80% |█████████████████████████▋      | 16.0MB 7.6MB/s eta 0:00:01
    100% |████████████████████████████████| 25.9MB 70kB/s  eta 0:00:011
    100% |████████████████████████████████| 307kB 2.2MB/s ta 0:00:01
You are using pip version 9.0.3, however version 21.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [3]:
#Importing required packages.
import pandas as pd
#import seaborn as sns
#import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import SGDClassifier
#from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score

final_results = []

In [4]:
#Loading dataset
wine = pd.read_csv('datasets/winequality-red.csv')

In [5]:
#Now seperate the dataset as response variable and feature variabes
X = wine.drop('quality', axis = 1)
y = wine['quality']

In [6]:
#Train and Test splitting of data 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [7]:
#Applying Standard scaling to get optimized result
sc = StandardScaler()

In [8]:
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [9]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(X_train, y_train)
print(lr.score(X_test, y_test))
final_results.append(lr.score(X_test, y_test))

0.40201256371442895


## Result should be 0.40201256371442895

## Preprocessing Data for performing Machine learning algorithms

In [10]:
#Making binary classificaion for the response variable.
#Dividing wine as good and bad by giving the limit for the quality
bins = (2, 6.5, 8)
group_names = ['bad', 'good']
wine['quality'] = pd.cut(wine['quality'], bins = bins, labels = group_names)

In [11]:
#Now lets assign a labels to our quality variable
label_quality = LabelEncoder()

In [12]:
#Bad becomes 0 and good becomes 1 
wine['quality'] = label_quality.fit_transform(wine['quality'])

In [13]:
#Now seperate the dataset as response variable and feature variabes
X = wine.drop('quality', axis = 1)
y = wine['quality']

In [14]:
#Train and Test splitting of data 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [15]:
#Applying Standard scaling to get optimized result
sc = StandardScaler()

In [16]:
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

### Random Forest Classifier

In [17]:
rfc = RandomForestClassifier(n_estimators=200, random_state=42)

rfc.fit(X_train, y_train)
print(rfc.score(X_test, y_test))
final_results.append(rfc.score(X_test, y_test))

0.8875


## Result should be 0.8875



## Stochastic Gradient Decent Classifier

In [18]:
sgd = SGDClassifier(penalty=None, random_state=42)
sgd.fit(X_train, y_train)
pred_sgd = sgd.predict(X_test)

In [19]:
print(sgd.score(X_test, y_test))
final_results.append(sgd.score(X_test, y_test))

0.846875


## Result should be 0.846875

## Support Vector Classifier

In [20]:
svc = SVC(random_state=42)
svc.fit(X_train, y_train)
pred_svc = svc.predict(X_test)

In [21]:
print(svc.score(X_test, y_test))
final_results.append(svc.score(X_test, y_test))

0.875


## Result should be 0.875

# GridSearchCV

In [22]:
#Finding best parameters for our SVC model
param = {
    'C': [0.1,0.8,0.9,1,1.1,1.2,1.3,1.4],
    'kernel':['linear', 'rbf'],
    'gamma' :[0.1,0.8,0.9,1,1.1,1.2,1.3,1.4]
}
grid_svc = GridSearchCV(svc, param_grid=param, scoring='accuracy', cv=10)

In [23]:
grid_svc.fit(X_train, y_train)

GridSearchCV(cv=10, estimator=SVC(random_state=42),
             param_grid={'C': [0.1, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4],
                         'gamma': [0.1, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4],
                         'kernel': ['linear', 'rbf']},
             scoring='accuracy')

In [24]:
#Best parameters for our svc model
grid_svc.best_params_

{'C': 1.2, 'gamma': 0.9, 'kernel': 'rbf'}

## Result should be {'C': 1.2, 'gamma': 0.9, 'kernel': 'rbf'}

In [25]:
#Let's run our SVC again with the best parameters.
svc2 = SVC(C = 1.2, gamma =  0.9, kernel= 'rbf', random_state=42)
svc2.fit(X_train, y_train)
pred_svc2 = svc2.predict(X_test)
#print(classification_report(y_test, pred_svc2))

In [26]:
print(svc2.score(X_test, y_test))
final_results.append(svc2.score(X_test, y_test))

0.896875


## Result should be 0.896875

## Cross Validation Score for random forest and SGD

In [27]:
#Now lets try to do some evaluation for random forest model using cross validation.
rfc_eval = cross_val_score(estimator = rfc, X = X_train, y = y_train, cv = 10)
print(rfc_eval.mean())
final_results.append(rfc_eval.mean())

0.9116818405511811


## Result should be 0.9116818405511811

In [28]:
print(final_results)

[0.40201256371442895, 0.8875, 0.846875, 0.875, 0.896875, 0.9116818405511811]
